daemon: NewDaemon: align grpc options with containerd's defaults

thaJeztah opened the pull request:
### daemon: NewDaemon: align grpc options with containerd's defaults

Dial-options passed to containerd _override_ all defaults that are set
in containerd, and containerd does not provide an option to provide
the defaults in other ways, which makes it slightly more complicated
to use the defaults combined with some custom options.
https://github.com/containerd/containerd/blob/v1.7.22/client.go#L122-L132

This patch aligns the options we set with the defaults in containerd.

grpc.FailOnNonTempDialError was added together with WithBlock in [containerd@64bc516],
but it looks like this was not copied to our options when the equivalent was
added in this repository through 9f73396dabf087a8dd5fa74296c2cd4c188ff889.

grpc.WithReturnConnectionError was added in [containerd@73d28dd] to improve
handling of connection errors;

Permission errors:

    % ./bin/ctr t ls
    ctr: failed to dial "/run/containerd/containerd.sock": connection error: desc = "transport: error while dialing: dial unix /run/containerd/containerd.sock: connect: permission denied"
    %

Non-existent sockets:

    % ./bin/ctr -a notfound t ls
    ctr: failed to dial "notfound": context deadline exceeded: connection error: desc = "transport: error while dialing: dial unix://notfound: timeout"
    %

That commit failed to notice that WithReturnConnectionError implies WithBlock,
so removing that option from the list.

Note that both WithBlock and WithReturnConnectionError are deprecated in
newer versions of grpc, so we should remove these once [containerd@63b4688]
makes it into the containerd 1.7 branch (and vendored).

[containerd@64bc516]: https://github.com/containerd/containerd/commit/64bc516bbef03e13ff09d9b40ec2f05f82bb387e
[containerd@73d28dd]: https://github.com/containerd/containerd/commit/73d28ddeb27a5532df084035199c9ad5ed2f69fe
[containerd@63b4688]: https://github.com/containerd/containerd/pull/10250/commits/63b46881753588624b2eac986660458318581330


### daemon: NewDaemon: remove grpc options that are the default

The default message size is set unconditionally in containerd's client,
so unlike Dial-options, there's no risk of implicitly dropping these
options.

TThis patch removes the options, as they were the same as the default
already set in containerd itself.

https://github.com/containerd/containerd/blob/v1.7.22/client.go#L133-L138

### libcontainerd/supervisor: remove grpc options that are the default

The default message size is set unconditionally in containerd's client,
so unlike Dial-options, there's no risk of implicitly dropping these
options.

TThis patch removes the options, as they were the same as the default
already set in containerd itself.

https://github.com/containerd/containerd/blob/v1.7.22/client.go#L133-L138



**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah committed:
daemon: NewDaemon: align grpc options with containerd's defaults

Dial-options passed to containerd _override_ all defaults that are set
in containerd, and containerd does not provide an option to provide
the defaults in other ways, which makes it slightly more complicated
to use the defaults combined with some custom options.
https://github.com/containerd/containerd/blob/v1.7.22/client.go#L122-L132

This patch aligns the options we set with the defaults in containerd.

grpc.FailOnNonTempDialError was added together with WithBlock in [containerd@64bc516],
but it looks like this was not copied to our options when the equivalent was
added in this repository through 9f73396dabf087a8dd5fa74296c2cd4c188ff889.

grpc.WithReturnConnectionError was added in [containerd@73d28dd] to improve
handling of connection errors;

Permission errors:

    % ./bin/ctr t ls
    ctr: failed to dial "/run/containerd/containerd.sock": connection error: desc = "transport: error while dialing: dial unix /run/containerd/containerd.sock: connect: permission denied"
    %

Non-existent sockets:

    % ./bin/ctr -a notfound t ls
    ctr: failed to dial "notfound": context deadline exceeded: connection error: desc = "transport: error while dialing: dial unix://notfound: timeout"
    %

That commit failed to notice that WithReturnConnectionError implies WithBlock,
so removing that option from the list.

Note that both WithBlock and WithReturnConnectionError are deprecated in
newer versions of grpc, so we should remove these once [containerd@63b4688]
makes it into the containerd 1.7 branch (and vendored).

[containerd@64bc516]: https://github.com/containerd/containerd/commit/64bc516bbef03e13ff09d9b40ec2f05f82bb387e
[containerd@73d28dd]: https://github.com/containerd/containerd/commit/73d28ddeb27a5532df084035199c9ad5ed2f69fe
[containerd@63b4688]: https://github.com/containerd/containerd/pull/10250/commits/63b46881753588624b2eac986660458318581330

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon: NewDaemon: remove grpc options that are the default

The default message size is set unconditionally in containerd's client,
so unlike Dial-options, there's no risk of implicitly dropping these
options.

TThis patch removes the options, as they were the same as the default
already set in containerd itself.

https://github.com/containerd/containerd/blob/v1.7.22/client.go#L133-L138

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

vvoland reviewed (approved):
LGTM (needs a rebase though)

thaJeztah committed:
libcontainer/supervisor: monitorDaemon: move options inline again

These were moved in 84965c075215de0add8968848e7f93219155be85

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

laurazard reviewed (approved):
LGTM

thaJeztah commented on daemon/daemon.go:
```diff
@@ -916,8 +915,14 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
+		// ------------------------------------------------------------------
+		// options below are copied from containerd client's default options
+		//
+		// TODO(thaJeztah): update this list once https://github.com/containerd/containerd/pull/10250/commits/63b46881753588624b2eac986660458318581330 is in the 1.7 release.
```
Oh! Maybe I can remove  more now, because we switched to the containerd v2 module; let me check before we merge.

thaJeztah committed:
libcontainerd/supervisor: remove grpc options that are the default

The default message size is set unconditionally in containerd's client,
so unlike Dial-options, there's no risk of implicitly dropping these
options.

TThis patch removes the options, as they were the same as the default
already set in containerd itself.

https://github.com/containerd/containerd/blob/v1.7.22/client.go#L133-L138

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon: NewDaemon: update grpc options for containerd 2.0

Now that we moved to use containerd 2.0, the changes from
https://github.com/containerd/containerd/commit/63b46881753588624b2eac986660458318581330
can now be used, removing some of gRPC's deprecated
options.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
daemon: NewDaemon: align max backoff delay with containerd 2.0

containerd 1.7 and lower set this value to a fixed, 3-second delay;
https://github.com/containerd/containerd/blob/v1.7.22/client.go#L117-L121

containerd 2.0 (starting with [containerd@63b4688]) aligned this value
with clientopts.timeout (WithTimeout()), which we set to 60 seconds;
https://github.com/containerd/containerd/blob/v2.0.2/client/client.go#L129-L136

This patch aligns our code with containerd client's defaults.

[containerd@63b4688]: https://github.com/containerd/containerd/commit/63b46881753588624b2eac986660458318581330

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented on daemon/daemon.go:
```diff
@@ -910,6 +910,9 @@ func NewDaemon(ctx context.Context, config *config.Config, pluginStore *plugin.S
+	// aligning with how containerd client's defaults sets this;
+	// https://github.com/containerd/containerd/blob/v2.0.2/client/client.go#L129-L136
 	backoffConfig := backoff.DefaultConfig
 	backoffConfig.MaxDelay = 3 * time.Second
```
done 👍
